dreadnode/agent/tools/fs.py

```diff
@@ -253,39 +284,85 @@ def grep(
                         for j in range(context_start, context_end):
                             prefix = ">" if j == i else " "
                             line_text = lines[j].rstrip("\r\n")
-                            context.append(f"{prefix} {j + 1}: {shorten_string(line_text, 80)}")
+                            context.append(
+                                f"{prefix} {j + 1}: {shorten_string(line_text, 80)}"
+                            )
 
                         rel_path = self._relative(file_path)
-                        matches.append(
+                        file_matches.append(
                             GrepMatch(
                                 path=rel_path,
                                 line_number=line_num,
                                 line=shorten_string(line.rstrip("\r\n"), 80),
                                 context=context,
                             ),
                         )
+            except (
+                FileNotFoundError,
+                PermissionError,
+                IsADirectoryError,
+                UnicodeDecodeError,
+                OSError,
+            ) as e:
+                logger.warning(f"Error occurred while searching file {file_path}: {e}")
```

Comment:
I assume we are trying to prevent isolated failures from blocking the whole tool call, just want to be thoughtful about the feedback to the model. Might be worth trying to see if we got only failures, then turn it into an error.

Reply:
We could also add a different data type to the returned objects like `list[GrepMatch | str]` and weave in error commentary.